### 01. Обзор данных - Wine Quality

В этом ноутбуке я делаю первое знакомство с датасетом.
Цель - понять, что за данные, какие столбцы, какого типа,
есть ли пропуски.

**Датасет:** `wine_quality_merged.csv` - объединённые данные
по красному и белому вину (6497 строк, 13 столбцов).

**Источник:** Kaggle (Red and White Wine Quality).

In [1]:
import pandas as pd


### Загрузка данных

Читаю CSV-файл в переменную `df` (от английского DataFrame - «таблица данных»).

`pd.read_csv()` - стандартная функция pandas для чтения CSV.

Путь `../data/raw/wine_quality_merged.csv` - потому что ноутбук лежит
в `notebooks/`, а CSV - в `data/raw/`. Две точки `..` значат «выйти на уровень выше».

In [2]:
df  = pd.read_csv("../data/raw/wine_quality_merged.csv")

### Сколько в таблице строк и столбцов?

`.shape` - это атрибут DataFrame. Возвращает кортеж (строки, столбцы).
Не функция - поэтому без скобок.

In [3]:
df.shape

(6497, 13)

### Названия столбцов

`.columns` - список названий столбцов. Тоже атрибут, без скобок.

In [4]:
df.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality', 'type'],
      dtype='str')

### Первые 5 строк

`.head()` - показывает первые 5 строк. Удобно, чтобы
увидеть, как данные выглядят «вживую».

In [5]:
df.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,type
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,red
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5,red
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5,red
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6,red
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,red


### Типы данных в столбцах

`.dtypes` - какие типы у каждого столбца.
`float64` - число с плавающей точкой (дробное),
`int64` - целое число,
`object` - обычно текст.

In [9]:
df.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
type                        str
dtype: object

### Базовая статистика по числовым столбцам

`.describe()` показывает для каждого числового столбца:
- count - сколько непустых значений
- mean - среднее
- std - стандартное отклонение (разброс)
- min, 25%, 50%, 75%, max - квантили 
> Квантили - это значения, которые делят отсортированный набор данных на четыре равные части.

Это быстрый способ увидеть, «как выглядят» числа:
есть ли странные значения, большой ли разброс и т.д.

In [10]:
df.describe()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
count,6497.000000,6497.000000,6497.000000,6497.000000,6497.000000,6497.000000,6497.000000,6497.000000,6497.000000,6497.000000,6497.000000,6497.000000
mean,7.215307,0.339666,0.318633,5.443235,0.056034,30.525319,115.744574,0.994697,3.218501,0.531268,10.491801,5.818378
std,1.296434,0.164636,0.145318,4.757804,0.035034,17.749400,56.521855,0.002999,0.160787,0.148806,1.192712,0.873255
min,3.800000,0.080000,0.000000,0.600000,0.009000,1.000000,6.000000,0.987110,2.720000,0.220000,8.000000,3.000000
25%,6.400000,0.230000,0.250000,1.800000,0.038000,17.000000,77.000000,0.992340,3.110000,0.430000,9.500000,5.000000
50%,7.000000,0.290000,0.310000,3.000000,0.047000,29.000000,118.000000,0.994890,3.210000,0.510000,10.300000,6.000000
75%,7.700000,0.400000,0.390000,8.100000,0.065000,41.000000,156.000000,0.996990,3.320000,0.600000,11.300000,6.000000
max,15.900000,1.580000,1.660000,65.800000,0.611000,289.000000,440.000000,1.038980,4.010000,2.000000,14.900000,9.000000


### Выводы по обзору данных

**Что узнал:**
- Датасет: 6497 строк, 13 столбцов.
- Пропусков нет (count везде = 6497).
- 11 числовых признаков + quality (int, целевой) + type (текст: red/white).

**Важные наблюдения:**
1. **Разные масштабы:** значения от 0.01 (chlorides) до 440 (total sulfur dioxide).
   При обучении KNN потребуется масштабирование (StandardScaler).
2. **Возможные выбросы:** residual sugar max = 65.8, тогда как 75% вин <= 8.1.
   Надо посмотреть на графике (boxplot / histogram).
3. **Несбалансированные классы в quality:** min=3, max=9, mean=5.82.
   Большинство оценок 5–6.

**Что делаем дальше:**
- Посмотрим распределение целевого признака quality.
- Сравним red и white по ключевым признакам.
- Построим графики (matplotlib/seaborn).